tests/thread_local/enc/enc.cpp

```diff
@@ -0,0 +1,147 @@
+
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+#include <openenclave/internal/tests.h>
+#include <stdio.h>
+#include <random>
+#include <set>
+#include <thread>
+
+#include "thread_local_t.h"
+
+// These variables will be put in .tdata
+__thread volatile int __thread_int = 1;
+__thread volatile int g_x[10] = {8};
+
+struct thread_local_struct
+{
+    bool initialized;
+    int value;
+
+    thread_local_struct(int v)
+    {
+        value = v;
+        initialized = true;
+        printf("thread_local_struct initialized with value = %d\n", value);
+    }
+    ~thread_local_struct()
+    {
+        printf("thread_local_struct destructed, value = %d\n", value);
+    }
+};
+
+// These variables will be put in .tbss
+// Dynamic thread specific destructors for g_s and dist.
+thread_local std::random_device g_rd;
+thread_local std::mt19937 g_mt(g_rd());
+thread_local std::uniform_real_distribution<double> g_dist(1, 100);
+thread_local volatile thread_local_struct g_s(g_dist(g_mt));
+
+// This variable will be put in .tdata
+thread_local int thread_local_int = 5;
+
+// Helper function for debugging.
+// Gets the value of the FS segment.
+void* get_fs()
+{
+    void* fs = NULL;
+    asm volatile("movq %%fs:0, %0" : "=r"(fs));
+    return fs;
+}
+
+// Helper class for spinlocks
+class spinlock
+{
+    std::atomic_flag locked = ATOMIC_FLAG_INIT;
+
+  public:
+    void acquire()
+    {
+        while (locked.test_and_set(std::memory_order_acquire))
```

Comment:
Shouldn't the test_and_set use memory_order_acq_rel because it is both synchronous read and write?

Reply:
Very interesting question. Got me thinking.
It only needs to be memory_order_acquire; doesn't have to be memory_order_acq_rel.

Let g_x = 0. Consider
(1)
g_x = 888;
flag.test_and_set(memory_order_acquire)

vs 
(2)
g_x = 888;
flag.test_and_set(memory_order_acq_rel)

The only difference is that since (1) is not a release operation, another thread could see the write to the flag and write to g_x in reverse order. So in another thread, the value of flag could be 1 while the value of g_x could still be 0.
Though the values are not up to date, it is fine since, in a properly written thread-safe code, the other thread is not supposed to access the (out-dated) value of g_x as it has not acquired the spinlock yet. It will go back to trying to acquire the lock.

When the current thread releases the spinlock via a memory_order_release, the other thread is guaranteed to see the write to g_x and the clearing of the flag in the correct order.

The only uses case for memory_order_acq_rel seems to be for transitive acquire-release semantics where 3 threads are involved.
int g_x = 0;
atomic_int flag = 0;


void thread1()
{
    // Update g_x value and wake up second thread.
    g_x = 888;
    flag.store(1, std::memory_order_release);
}

void thread2()
{
    // Wait for flag to be 1 and atomically set it to 2
   int expected=1;
    while (!flag.compare_exchange_strong(expected, 2, std::memory_order_acq_rel)) {
        expected = 1;
    }
    // Since this thread uses acquire, it will see the latest value of g_x and the flag in the  correct order.
   //  That means in this thread the value of g_x has been updated or written to 888.
   //  Even though the first thread actually wrote to g_x, from the 3rd thread's perspective it is as if this thread
   //  wrote to g_x.
   // If this thread uses a simple write to the flag (ie memory_order_relaxed), then the write to g_x and flag
   // in the thread will not be seen in correct order by the next thread.
   // Therefore this thread has to use memory_order_release to write to the flag so that the next thread sees the
   // write to g_x and flag in correct order.
}

void thread3()
{
    // Wait for flag to be 2
    while (flag.load(std::memory_order_acquire) < 2) ;
    // If the second thread uses release semantics, this assertion will never fail.
    assert(g_x == 888);
}

Based on second example from: https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering